src/Resolver/Location/FindParentStrategy.php

```diff
@@ -129,7 +136,28 @@ public function updateParent(ElementInterface $element, array $inputData): Eleme
         }
 
         if ($newParent) {
-            return $element->setParent($newParent);
+            if ($newParent->getType() === AbstractObject::OBJECT_TYPE_VARIANT) {
```

Comment:
in UI we allow variants of variants, so I think we should not restrict that here either.

Reply:
Right, I did overlook this. So I now added the condition, that a variant can be set as a parent of a variant, if it's of the same class as the element itself.

Reply:
sorry for being picky, but that condition actually applies for all parents .. a variant always has to be of the same type as its parent, regardless if it is an object or a variant.

Reply:
It's fine if you're being picky, I actually appreciate that 😃 

But I'm not entirely sure, if I get you right here. A variant needs to be of the same class as its parent - this is what I actually want to verify here. Let me try elaborate what my intention was - and you can - of course - correct me if I'm wrong.

I didn't want it to be possible to set a parent to any object, if that parent is a variant, unless the element itself is a variant and it's of the same class as the parent. Otherwise you could assign non variant elements to a parent that is a variant, or assign variants to a variant of another class.

https://github.com/pimcore/data-importer/blob/830fef0deb5bb29098e6ebee2b7394a034ab671a/src/Resolver/Location/FindParentStrategy.php#L139-L149

First, I'm checking if the parent is a variant. If this is the case, the element needs to be a variant of the same class as well. Otherwise, I could either assign a variant of another class - or any other class types to a variant - which is also not possible in the admin-ui.

Am I missing something? Or is there already a built in functionality checking this?
If you think the import should allow that, I'm also totally fine with that and I will remove the entire condition.


---

Another thing I just realized is that the following condition also is not working for variants having variants:
https://github.com/pimcore/data-importer/blob/830fef0deb5bb29098e6ebee2b7394a034ab671a/src/Resolver/Location/FindParentStrategy.php#L157
I'm biased on this one. We have the following possibilities:
1. you can only change an object to a variant if it hasn't set any children yet, this also has the advantage that it won't be possible to create a tree structure where no parent of type object exists.
1. we check for all children, if they are variants of the object - that sounds a little performance hungry and imo should be avoided
1. or we could ignore that and simply make it possible to change an object to a variant regardless if it has any children or not

I actually prefer the current solution (1)

Reply:
I get your point. But there is still an inconsistency. 
1) Trying to assign a variant of different class below a variant leads to an exception, because of https://github.com/pimcore/data-importer/pull/375/files#diff-8f9965171cbd95a20bc79746f27b5472c3243e89c9b0e865fdc2618168c25a39R145-R149
2) Trying to assign a variant of different class below an object is just silently omitted by not creating a variant but an normal object, because of https://github.com/pimcore/data-importer/pull/375/files#diff-8f9965171cbd95a20bc79746f27b5472c3243e89c9b0e865fdc2618168c25a39R158-R164

Or do I miss something? 

I would say, we should throw exceptions in all cases when creating/updating a variant is not possible but configured.

Reply:
why is `!$element->hasChildren() ` causing problems with variants having variants? 

but I'm with you, we should allow converting to variants only for objects without any children (what this condition actually does, right?).

Reply:
> why is `!$element->hasChildren() ` causing problems with variants having variants?
> 

I was thinking about the case where you have
```
object (1)
|- object (2)
|-- object (3)
```
with the `!$element->hasChildren() ` check you cannot change `object (2)` to a variant, although it might be legit.
```
object (1)
|- variant (2) <- this doesn't work, as it has a child
|-- object (3)
```

> but I'm with you, we should allow converting to variants only for objects without any children (what this condition actually does, right?).

Exactly. So we'll leave it like that.

Reply:
>I would say, we should throw exceptions in all cases when creating/updating a variant is not possible but configured.

Ok, I will create an exception for any case